Added falloff values for Heal, Attack & Repair.

ViezeVingertjes opened the pull request:
Correct me if i'm wrong, i was not sure which constant to use but RANGED_HEAL_POWER made sense for towers.

TheRealMaxion commented:
TOWER_POWER_HEAL: 400

TheRealMaxion commented:
And in combat you want the overhealing.

anisoptera commented:
Agree, use the constant and add `// TODO: combat overhealing` :)

TheRealMaxion commented:
The actual healing power is also different with range

ViezeVingertjes commented:
So many variables haha... could you point me to the docs where i can find information about the range dropoff?

TheRealMaxion commented:
TOWER_OPTIMAL_RANGE: 5,
TOWER_FALLOFF_RANGE: 20,
TOWER_FALLOFF: 0.75
 At range 5 or less it has full power, at range 20 or more it has (1 - 0.75) times less power, and it scales linearly in between.

ViezeVingertjes commented:
Replaced the static values with a more dynamic (future proof) version. Also renamed the pull request to reflect the current feature. Only Heal is used for now, but at almost no extra cost the attack & repair values are added to be used in the future.

tedivm commented on src/programs/city/defense.js:
```diff
@@ -8,6 +8,19 @@ class CityDefense extends kernel.process {
+        global.TOWER_REPAIR_EFFECT[i] = this.calculateWithFallOff(TOWER_POWER_REPAIR, i)
+        global.TOWER_HEAL_EFFECT[i] = this.calculateWithFallOff(TOWER_POWER_HEAL, i)
+      }
+    }
```
I think these should be defined in `constants` so they can be reused throughout the code.

tedivm replied:
They should also get placed in `package.json` so the linter won't toss an error.

ButAds commented on src/programs/city/defense.js:
```diff
@@ -55,27 +68,36 @@ class CityDefense extends kernel.process {
+        }
+
+        const distance = tower.pos.getRangeTo(healTarget.pos)
+        damage -= global.TOWER_HEAL_EFFECT[distance]
```
might break if the target, for whatever reason, is in another room. I'd suggest clamping the range to the max room size

ButAds commented on src/programs/city/defense.js:
```diff
@@ -131,6 +153,20 @@ class CityDefense extends kernel.process {
+  calculateWithFallOff (optimalValue, distance) {
+    let factor = 0.0
+
+    if (distance < TOWER_OPTIMAL_RANGE) {
```
Can also be written as:
```javascript
let effect = optimalValue;
if(distance > TOWER_OPTIMAL_RANGE) {
    if(distance> TOWER_FALLOFF_RANGE) {
        distance= TOWER_FALLOFF_RANGE;
    }
    effect -= effect * TOWER_FALLOFF * (distance - TOWER_OPTIMAL_RANGE) / (TOWER_FALLOFF_RANGE - TOWER_OPTIMAL_RANGE);
}
return Math.floor(effect);
```
Related to the calculation code in the engine
https://github.com/screeps/engine/blob/master/src/processor/intents/towers/attack.js#L34
